File: language/verus/examples/binary_search.rs
use vstd::prelude::*;

verus! {

fn binary_search(a: &Vec<i64>, target: i64) -> (index: usize)
    requires
        forall|i: int, j: int| 0 <= i < j < a.len() ==> a[i] <= a[j],
    ensures
        index < a.len() ==> a[index as int] == target,
        index == a.len() ==> forall|k: int| 0 <= k < a.len() ==> a[k] != target,
{
    let mut lo: usize = 0;
    let mut hi: usize = a.len();

    while lo < hi
        invariant
            0 <= lo <= hi <= a.len(),
            forall|k: int| 0 <= k < lo as int ==> a[k] != target,
            forall|k: int| hi as int <= k < a.len() ==> a[k] != target,
            forall|i: int, j: int| 0 <= i < j < a.len() ==> a[i] <= a[j],
        decreases hi - lo,
    {
        let mid = lo + (hi - lo) / 2;
        if a[mid] == target {
            return mid;
        } else if a[mid] < target {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    a.len()
}

} // verus!

File: language/framac/examples/binary_search.c
/*@
  requires n > 0;
  requires \valid(a + (0 .. n-1));
  requires \forall integer i, j; 0 <= i <= j < n ==> a[i] <= a[j];
  assigns \nothing;
  ensures -1 <= \result < n;
  ensures \result >= 0 ==> a[\result] == v;
  ensures \result == -1 ==> \forall integer i; 0 <= i < n ==> a[i] != v;
*/
int binary_search(int *a, int n, int v) {
  int lo = 0, hi = n;
  /*@
    loop invariant 0 <= lo <= hi <= n;
    loop invariant \forall integer i; 0 <= i < lo ==> a[i] != v;
    loop invariant \forall integer i; hi <= i < n ==> a[i] != v;
    loop assigns lo, hi;
    loop variant hi - lo;
  */
  while (lo < hi) {
    int mid = lo + (hi - lo) / 2;
    if (a[mid] == v) return mid;
    else if (a[mid] < v) lo = mid + 1;
    else hi = mid;
  }
  return -1;
}
